# Empirical Investigation: Statistical Test

pandas is a powerful library for data manipulation and analysis.

statsmodels.stats.weightstats is a module in the statsmodels library that provides functions for conducting hypothesis tests, such as the independent samples t-test.

csv is a library for reading and writing CSV files.






In [1]:
import pandas as pd
from statsmodels.stats.weightstats import ttest_ind
import csv

In [2]:
with open('final_accs.csv', 'r', encoding='utf-8') as file:
    reader = csv.reader(file)
    header = next(reader)
    data = [row for row in reader]

df = pd.DataFrame(data, columns=header)

The code below opens the CSV file named final_accs.csv and reads its contents into a csv.reader object. The header variable is set to the first row of the file, which contains the names of the columns. The remaining rows are stored in the data variable as a list of lists, where each inner list represents a row in the file.

Finally, the data is transformed into a Pandas DataFrame using the pd.DataFrame function. The data list is passed as the first argument, and header is passed as the columns argument, which specifies the column names for the DataFrame.

At this point, we have successfully loaded the data from the CSV file into a Pandas DataFrame, which makes it easy to work with and manipulate the data.

In [3]:
df = df.astype(float)


In [4]:
df_runtime = df[['ff_train_time', 'ff_test_time', 'fb_train_time', 'fb_test_time']]
df_runtime

ff_train_time  ff_test_time  fb_train_time  fb_test_time
0      106.953362      2.454056       4.020643      0.017794
1      106.817223      1.951378       4.012678      0.018571
2      106.863541      2.411181       4.040795      0.017637
3      107.069862      1.963288       4.019366      0.018249
4      107.060204      2.325379       4.039193      0.017631
5      106.904060      1.990357       4.055216      0.017784
6      107.002090      1.989709       4.028404      0.017763
7      106.978630      2.016014       4.020052      0.012162
8      106.884392      1.992291       4.040663      0.017647
9      106.964607      1.896359       4.013278      0.017724
10     107.031555      1.842056       4.019631      0.017527

In [5]:
#vergleich mit Ausgabe
mean_values = df_runtime.mean()
print(mean_values)

ff_train_time    106.957230
ff_test_time       2.075643
fb_train_time      4.028175
fb_test_time       0.017317
dtype: float64


The code calculates the mean of each column in the DataFrame and stores the result in a variable named mean_values. The mean function from the DataFrame is called, which calculates the mean of each column.

Finally, the result is printed to the console using the print function. The mean of each column is displayed, allowing us to inspect the values and verify that the mean calculation was performed correctly.

In [6]:
df_accuracy = df[['ff_train_acc', 'ff_test_acc', 'fb_train_acc', 'fb_test_acc']]
df_accuracy

ff_train_acc  ff_test_acc  fb_train_acc  fb_test_acc
0        0.94870       0.9494       0.99143       0.9736
1        0.94768       0.9480       0.99122       0.9739
2        0.94838       0.9494       0.99060       0.9733
3        0.94800       0.9495       0.99101       0.9714
4        0.94920       0.9510       0.99060       0.9719
5        0.94848       0.9473       0.99067       0.9716
6        0.94622       0.9472       0.99155       0.9754
7        0.94736       0.9493       0.99164       0.9722
8        0.94832       0.9476       0.99157       0.9742
9        0.94600       0.9447       0.99120       0.9723
10       0.94848       0.9500       0.98998       0.9719

In [7]:
mean_values_acc = df_accuracy.mean()
print(mean_values_acc)

ff_train_acc    0.947893
ff_test_acc     0.948491
fb_train_acc    0.991043
fb_test_acc     0.972882
dtype: float64


The code calculates the mean of each column in a DataFrame named df_accuracy and stores the result in a variable named mean_values_acc. The mean function from the DataFrame is called, which calculates the mean of each column.

Finally, the result is printed to the console using the print function. The mean of each column is displayed, allowing us to inspect the values and verify that the mean calculation was performed correctly.

In [ ]:
# Perform the t-test and get the p-value on the train data
p = ttest_ind(df_accuracy['ff_train_acc'], df_accuracy['fb_train_acc'])
print("p-value:", p)

p-value: (-125.75562264122729, 1.8218429834622805e-30, 20.0)


The p-value in this case appears to be a tuple of three values: -125.75562264122729, 1.8218429834622805e-30, and 20.0. These values can be interpreted as follows:

-125.75562264122729: This value is the calculated test statistic for the hypothesis test, which measures the difference between the two sample means and provides a measure of how far apart the two means are from each other. A negative value of the test statistic indicates that the first mean is lower than the second mean, while a positive value indicates that the first mean is higher than the second mean.

1.8218429834622805e-30: This value is the p-value, which is the probability of obtaining a test statistic as extreme or more extreme than the one observed, assuming that the null hypothesis is true. A small p-value, such as the one observed in this case, suggests that the null hypothesis is unlikely to be true, and provides strong evidence for rejecting the null hypothesis in favor of the alternative hypothesis.

20.0: This value could be the degrees of freedom for the t-test, which is equal to the total number of observations in both samples minus the number of parameters estimated from the sample data. The degrees of freedom determine the shape of the t-distribution used to calculate the p-value.

In this case, the p-value of 1.8218429834622805e-30 is extremely small and suggests that there is strong evidence against the null hypothesis. The researcher can therefore reject the null hypothesis and conclude that there is a statistically significant difference between the two sample means.






In [ ]:
# Perform the t-test and get the p-value on the test data
p = ttest_ind(df_accuracy['ff_test_acc'], df_accuracy['fb_test_acc'])
print("p-value:", p)

p-value: (-37.407938593481234, 5.472573092936269e-20, 20.0)


The p-value in this case appears to be a tuple of three values: -37.407938593481234, 5.472573092936269e-20, and 20.0. These values can be interpreted as follows:

-37.407938593481234: This value is the calculated test statistic for the hypothesis test, which measures the difference between the two sample means and provides a measure of how far apart the two means are from each other. A negative value of the test statistic indicates that the first mean is lower than the second mean, while a positive value indicates that the first mean is higher than the second mean.

5.472573092936269e-20: This value is the p-value, which is the probability of obtaining a test statistic as extreme or more extreme than the one observed, assuming that the null hypothesis is true. A small p-value, such as the one observed in this case, suggests that the null hypothesis is unlikely to be true, and provides strong evidence for rejecting the null hypothesis in favor of the alternative hypothesis.

20.0: This value could be the degrees of freedom for the t-test, which is equal to the total number of observations in both samples minus the number of parameters estimated from the sample data. The degrees of freedom determine the shape of the t-distribution used to calculate the p-value.

In this case, the p-value of 5.472573092936269e-20 is extremely small and suggests that there is strong evidence against the null hypothesis. The researcher can therefore reject the null hypothesis and conclude that there is a statistically significant difference between the two sample means.

# Visualisierung

In [35]:
import plotly.express as px
import requests

In [37]:
import plotly.graph_objects as go
import pandas as pd

df = pd.DataFrame({
    'ff_train_acc': [0.94870, 0.94768, 0.94838, 0.94800, 0.94920, 0.94848, 0.94622, 0.94736, 0.94832, 0.94600, 0.94848],
    'ff_test_acc': [0.9494, 0.9480, 0.9494, 0.9495, 0.9510, 0.9473, 0.9472, 0.9493, 0.9476, 0.9447, 0.9500],
    'fb_train_acc': [0.99143, 0.99122, 0.99060, 0.99101, 0.99060, 0.99067, 0.99155, 0.99164, 0.99157, 0.99120, 0.98998],
    'fb_test_acc': [0.9736, 0.9739, 0.9733, 0.9714, 0.9719, 0.9716, 0.9754, 0.9722, 0.9742, 0.9723, 0.9719]
})

fig = go.Figure()
for col in df.columns:
    fig.add_trace(go.Box(y=df[col], name=col, boxpoints='all', jitter=1, pointpos=-2))

fig.update_layout(title='Model Performance',
                  xaxis=dict(title='Model'),
                  yaxis=dict(title='Accuracy'))
fig.show()


In [26]:
import plotly.graph_objects as go
import pandas as pd

df = pd.DataFrame({
    'ff_train_time': [106.953362, 106.817223, 106.863541, 107.069862, 107.060204, 106.904060, 107.002090, 106.978630, 106.884392, 106.964607, 107.031555],
    'ff_test_time': [2.454056, 1.951378, 2.411181, 1.963288, 2.325379, 1.990357, 1.989709, 2.016014, 1.992291, 1.896359, 1.842056],
    'fb_train_time': [4.020643, 4.012678, 4.040795, 4.019366, 4.039193, 4.055216, 4.028404, 4.020052, 4.040663, 4.013278, 4.019631],
    'fb_test_time': [0.017794, 0.018571, 0.017637, 0.018249, 0.017631, 0.017784, 0.017763, 0.012162, 0.017647, 0.017724, 0.017527]
})

fig = go.Figure()
for col in df.columns:
    fig.add_trace(go.Box(y=df[col], name=col, boxpoints='all', jitter=1, pointpos=-2))

fig.update_layout(title='Model Performance',
                  xaxis=dict(title='Model'),
                  yaxis=dict(title='Runtime'))
fig.show()